### ISSUE TITLE
Gorouter retry-logic no longer works correctly when route integrity is enabled

### ISSUE THREAD

**Issue by domdom82** [cloudfoundry/routing-release#187]
## Issue
I've observed that broken app instances still get requests by gorouter, even though the logic says it should retry another instance.

## Affected Versions
Latest up to 0.208.0

## Context
A customer complaint made me test out how the retry logic in [round_tripper](https://github.com/cloudfoundry/gorouter/blob/main/proxy/round_tripper/proxy_round_tripper.go#L127) actually works:

1. Loadbalancer selects next endpoint (roundtrip or least conn)
2. Try round trip
3. On error, check if error is retryable (retriable?)
4. If can be retried, pick next instance

"retryable" here means the error is member of the exclusive [retriables group](https://github.com/cloudfoundry/gorouter/blob/main/proxy/fails/classifier_group.go#L13) where only very basic TCP and TLS errors can join.

In the pre-TLS era gorouter would always directly talk to the app instance so if the app failed to accept the request, it would report an error such as "connection refused" or "dial timeout".

However, since TLS has been enabled, for gorouter the de facto endpoint has become envoy, not the app. My suspicion is that since envoy will always at least accept a TCP connection and a TLS handshake, the above retry-logic fails to apply.

## Steps to Reproduce

If my assumptions are correct a broken app should only ever lead to at most one 502 error (at the time it breaks during a request). If there are more than 1 instance of the app, gorouter should use the healthy instance on the next request as it failed to connect to the broken one.

To test this behavior, I wrote a [small app](https://github.com/domdom82/go-fail) that breaks on purpose.

**Scenario:**

- App fails mid-request (instance dies)
- Gorouter must return 502
- Next request, gorouter **should** not be able to talk to app instance
- Gorouter **should** select other app instance
- Only one request should fail

**Test:**

- deploy app that fails on demand
- target specific instance of app and kill it
- issue a number of requests against the app

**Test Sequence:**
- 5 normal requests
- 1 request to kill instance
- 5 normal requests

## Expected result

- only the "kill" request should fail

## Current result

- several other requests fail. All of them target the broken instance.
- this goes on until eventually the route to the broken app is removed from NATS

## Possible Fix

I haven't dug into the actual error returned by envoy yet. It is possible there is just another case we haven't yet covered in the "retriables group". However, I fear it is not so trivial and there might be no error at all returned by envoy. 

In that case it might be an idea to mark an endpoint as "temporarily unavailable" after a 502 has been reported for it. This would buy NATS enough time to remove the crashed instance from the routing table. This could be done by putting a timestamp on the routing table entry. The next time that endpoint is selected (and other endpoints are available) we could check the timestamp and pick another one if not enough time (e.g. 5 seconds) has passed since we last tried it.

## Additional Context

Here are my test runs:

(the app prints the instance index as well a random number 0-100 and crashes the instance if the given probability (p) is lower)
(there are two instances, the test will always crash instance with index 1)

### No TLS (no envoy):
```
run 1:
(instance: 1) 43 >= 0 - OK
(instance: 0) 35 >= 0 - OK
(instance: 1) 13 >= 0 - OK
(instance: 0) 21 >= 0 - OK
(instance: 1) 24 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 0 >= 0 - OK
(instance: 0) 69 >= 0 - OK
(instance: 0) 11 >= 0 - OK
(instance: 0) 79 >= 0 - OK
(instance: 0) 9 >= 0 - OK

run 2:
(instance: 1) 49 >= 0 - OK
(instance: 0) 87 >= 0 - OK
(instance: 1) 72 >= 0 - OK
(instance: 0) 10 >= 0 - OK
(instance: 1) 15 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 9 >= 0 - OK
(instance: 0) 80 >= 0 - OK
(instance: 0) 98 >= 0 - OK
(instance: 0) 47 >= 0 - OK
(instance: 0) 70 >= 0 - OK

run 3:
(instance: 0) 3 >= 0 - OK
(instance: 1) 83 >= 0 - OK
(instance: 0) 63 >= 0 - OK
(instance: 1) 40 >= 0 - OK
(instance: 0) 20 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 49 >= 0 - OK
(instance: 0) 79 >= 0 - OK
(instance: 0) 11 >= 0 - OK
(instance: 0) 72 >= 0 - OK
(instance: 0) 20 >= 0 - OK
```

### With mTLS (use envoy):
```
run 1:
(instance: 0) 64 >= 0 - OK
(instance: 1) 71 >= 0 - OK
(instance: 0) 38 >= 0 - OK
(instance: 1) 95 >= 0 - OK
(instance: 0) 13 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 14 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 67 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.

run 2:
(instance: 1) 67 >= 0 - OK
(instance: 0) 68 >= 0 - OK
(instance: 1) 34 >= 0 - OK
(instance: 0) 13 >= 0 - OK
(instance: 1) 91 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 40 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 76 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 98 >= 0 - OK

run 3:
(instance: 0) 83 >= 0 - OK
(instance: 1) 7 >= 0 - OK
(instance: 0) 24 >= 0 - OK
(instance: 1) 83 >= 0 - OK
(instance: 0) 55 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 13 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
(instance: 0) 11 >= 0 - OK
502 Bad Gateway: Registered endpoint failed to handle the request.
```

### Here is the script that runs the test
```
#!/usr/bin/env bash

APP_ID="847da9d3-edf5-4bf6-a41e-c1a462ad18b0"
APP_INDEX="1"

curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0

curl -H "X-Cf-App-Instance:$APP_ID:$APP_INDEX" https://go-fail.cfapps.mycloud.com/fail?p=100

curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
curl https://go-fail.cfapps.mycloud.com/fail?p=0
```

**Comment by ameowlia** [comment]
Hi @domdom82 , 

Thanks for writing this up and especially for making a nice little testing app. We will try to reproduce soon and then we can discuss possible solutions with you.

Thanks, 
Amelia

**Comment by domdom82** [comment]
Hi @ameowlia, I've had another look at this today. Unfortunately, the returned error is just `io.EOF` which is not classified as retry-able nor fail-able in the `ClassifierGroup` so it won't be retried nor marked as failed.
I've prepared some code that adds it as a fail-able error since it would be dangerous to retry it (could have been a post request).

However I am not really happy with this kind of error, it could be anything. Also `io.EOF` does not necessarily mean it was an error at all, it could also have been emitted by the app itself after it was done writing to a stream.

So before I continue with my approach I wonder if there is anything in the Envoy config we could change so that a crashed app makes Envoy refuse a connection or emit a better error. 

I've downloaded the Envoy config for my test app and noticed there are no health checks configured for it:
```
(...)
   "static_clusters": [
    {
     "cluster": {
      "@type": "type.googleapis.com/envoy.api.v2.Cluster",
      "name": "0-service-cluster",
      "type": "STATIC",
      "connect_timeout": "0.250s",
      "circuit_breakers": {
       "thresholds": [
        {
         "max_connections": 4294967295
        }
       ]
      },
      "load_assignment": {
       "cluster_name": "0-service-cluster",
       "endpoints": [
        {
         "lb_endpoints": [
          {
           "endpoint": {
            "address": {
             "socket_address": {
              "address": "10.137.125.171",
              "port_value": 8080
             }
            }
           }
          }
         ]
        }
       ]
      }
     }
(...)
```

So from Envoy's perspective, the app is always healthy and it will try to direct traffic to it. My hope is that if we introduce a health check and Envoy removes the broken app, thus creating an empty cluster, it won't forward the next request and instead reject it so that Gorouter would get a better error. Haven't tested it yet :-)

**Comment by peterellisjones** [comment]
Hi folks, with @domdom82's help I did some testing with Envoy today which unfortunately seems to suggest that there is no healthcheck configuration for Envoy that would cause Envoy to only accept an incoming TCP connection if the application is healthy.

I deployed Envoy and a test app locally, set up an Envoy healthcheck for the app so that it _always fails_ and configured Envoy to disable  [panic mode](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/panic_threshold#arch-overview-load-balancing-panic-threshold) to ensure that requests are never sent to unhealthy instances. The Envoy admin cluster status reports that Envoy considers the cluster unhealthy, and the docker compose logs confirm that Envoy is not forwarding the request to the test app. When curling I get “Empty reply from server” which from everything I’ve read means that a TCP connection was established but no HTTP payload was transmitted (netcat also confirms that a TCP connection is established with Envoy). Repeating the experiment with go’s net/http library gives an EOF error which I expect means the same thing. Unfortunately as this seems to be indistinguishable from the case where a request is received by the application but where the application choses not to write an http response, I don’t think we can retry on this error without risking [mutating non-idempotent requests being retried](https://github.com/cloudfoundry/gorouter/blob/main/proxy/fails/classifier_group.go#L11-L12).

To summarise the issue
* The only classes of error that are safe to retry are ones where a TCP connection is not successfully established
* There does not seem to be a configuration for Envoy that makes it reject/ignore TCP connections when the application is unhealthy, it seems to always accept a connection even if it has no intention of proxying the request due to the cluster being unhealthy

**Comment by peterellisjones** [comment]
Note that this issue means that the [docs](https://docs.cloudfoundry.org/concepts/http-routing.html#retry) for gorouter transparent retries are not quite accurate when route integrity is enabled

![Screenshot 2021-05-19 at 10 32 34](https://user-images.githubusercontent.com/2788556/118790330-91be4f00-b88d-11eb-9961-c9beff0dc777.png)

With route integrity enabled, Envoy accepts a TCP connection even if the app is not listening, which means that in this situation the user will get a 502 rather than the gorouter retrying another instance. I'm not sure what the Envoy lifecycle looks like compared to the app lifecycle but I expect there is a short period of time when a Diego cell is recreated or app instances are migrated where Envoy is listening but the app is not.

**Comment by domdom82** [comment]
@peterellisjones With your PR merged let's try to reproduce this issue. If my understanding is correct, a failing app should at most produce one 502 before gorouter will retry another instance. If that is the case, the old behavior is restored and we can close this issue.

**Comment by peterellisjones** [comment]
I've manually confirmed that [routing release 0.217.0](https://github.com/cloudfoundry/routing-release/releases/tag/0.217.0) fixes this issue -- all subsequent requests to the crashed app are correctly retried against the healthy instance:

With 0.216.0:
```
> ./test.sh
(instance: 1) 64 >= 0 - OK
(instance: 0) 16 >= 0 - OK
(instance: 1) 76 >= 0 - OK
(instance: 1) 28 >= 0 - OK
(instance: 0) 55 >= 0 - OK
curl: (22) The requested URL returned error: 502 Bad Gateway
curl: (22) The requested URL returned error: 502 Bad Gateway
curl: (22) The requested URL returned error: 502 Bad Gateway
curl: (22) The requested URL returned error: 502 Bad Gateway
(instance: 0) 71 >= 0 - OK
(instance: 0) 68 >= 0 - OK
```

With 0.217.0:
```
> ./test.sh
(instance: 0) 95 >= 0 - OK
(instance: 0) 41 >= 0 - OK
(instance: 1) 2 >= 0 - OK
(instance: 1) 69 >= 0 - OK
(instance: 0) 6 >= 0 - OK
curl: (22) The requested URL returned error: 502 Bad Gateway
(instance: 0) 33 >= 0 - OK
(instance: 0) 8 >= 0 - OK
(instance: 0) 96 >= 0 - OK
(instance: 0) 50 >= 0 - OK
(instance: 0) 70 >= 0 - OK
```

**Comment by domdom82** [comment]
@peterellisjones Excellent! Closing.
